rust/lance/src/dataset.rs

```diff
@@ -401,10 +418,80 @@ impl Dataset {
         Ok(())
     }
 
-    fn already_checked_out(&self, location: &ManifestLocation) -> bool {
+    /// Check out the latest version of the branch
+    pub async fn checkout_branch(&self, branch: &str) -> Result<Self> {
+        self.checkout_by_ref(None, Some(branch.to_string())).await
+    }
+
+    /// This is a two-phase operation:
+    /// - Create the branch dataset by shallow cloning.
+    /// - Create the branch metadata (a.k.a. `BranchContent`).
+    ///
+    /// These two phases are not atomic. We consider `BranchContent` as the source of truth
+    /// for the branch.
+    ///
+    /// The cleanup procedure should:
+    /// - Clean up zombie branch datasets that have no related `BranchContent`.
+    /// - Delete broken `BranchContent` entries that have no related branch dataset.
+    ///
+    /// If `create_branch` stops at phase 1, it may leave a zombie branch dataset,
+    /// which can be cleaned up later. Such a zombie dataset may cause a branch creation
+    /// failure if we use the same name to `create_branch`. In that case, you need to call
+    /// `force_delete_branch` to interactively clean up the zombie dataset.
+    pub async fn create_branch(
+        &mut self,
+        branch: &str,
+        version: impl Into<refs::Ref>,
+        store_params: Option<ObjectStoreParams>,
+    ) -> Result<Self> {
+        let (source_branch, version_number) = self.resolve_reference(version.into()).await?;
+        let branch_location = self.find_branch_location(branch)?;
+        let clone_op = Operation::Clone {
+            is_shallow: true,
+            ref_name: source_branch.clone(),
+            ref_version: version_number,
+            ref_path: String::from(self.uri()),
+            branch_name: Some(branch.to_string()),
+        };
+        let transaction = Transaction::new(version_number, clone_op, None, None);
+
+        let builder = CommitBuilder::new(WriteDestination::Uri(branch_location.uri.as_str()))
+            .with_store_params(store_params.unwrap_or_default())
+            .with_object_store(Arc::new(self.object_store().clone()))
+            .with_commit_handler(self.commit_handler.clone())
+            .with_storage_format(self.manifest.data_storage_format.lance_file_version()?);
+        let dataset = builder.execute(transaction).await?;
+
+        // Create BranchContent after shallow_clone
```

Comment:
in various places across the code, should use `BranchContents` instead of `BranchContent`